File: leetcode/python/143. 重排链表.py
#!/bin/bash/python
"""
143. 重排链表

思路一：
递归（超时）

思路二：
使用list保存所有节点
依次从首尾添加节点

思路三：
找到中间节点
把后半部分保存到list
依次插入到前半部分中

思路四：
找到中间节点
后半部分翻转链表
两个子链表从头归并
"""

class ListNode(object):
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution(object):
    def reorderList(self, head):
        """
        递归法
        """
        if head == None or head.next == None or head.next.next == None:
            return head
        # 找到当前的非空尾节点和上一个节点
        p, q = head, head.next
        while q.next:
            p = p.next
            q = q.next
        # 重新排列
        p.next = None
        newHead = head.next
        head.next = q
        q.next = self.reorderList(newHead)
        return head

    def reorderList2(self, head):
        """
        全部保存法
        """
        # 使用list保留所有的节点
        stack = []
        tmp = head
        while tmp:
            stack.append(tmp)
            tmp = tmp.next
        # 分别从前后取值
        newHead = ListNode()
        tmp = newHead
        index = 0
        while stack:
            tmp.next = stack.pop(index)
            tmp = tmp.next
            index = -1 - index
        tmp.next = None
        return newHead.next

    def reorderList3(self, head):
        """
        后一半保存，插入前一半
        好像不用返回值
        """
        if head == None:
            return None
        # 获取中间节点
        p = q = head
        while p and q and q.next:
            p = p.next
            q = q.next.next
        # 保存后半部分到栈
        stack = []
        while p:
            stack.append(p)
            p = p.next
        # 从头依次完成插入操作
        while head and len(stack) > 1:
            tmp = head.next
            head.next = stack.pop()
            head.next.next = tmp
            head = tmp
            if stack:
                stack[-1].next = None
    
    def reorderList4(self, head):
        """
        找中点，翻转，归并
        """
        if head == None and head.next == None:
            return
        # 找中点，断开
        mid = p = q = head
        while q and q.next:
            mid = p
            p = p.next
            q = q.next.next
        mid.next = None
        # 后半部分翻转
        newHead = None
        while p:
            tmp = p.next
            p.next = newHead
            newHead = p
            p = tmp
        # 两链表合并
        h = ListNode()
        while head or newHead:
            if head:
                h.next = head
                h = h.next
                head = head.next
            if newHead:
                h.next = newHead
                h = h.next
                newHead = newHead.next

File: leetcode/cpp/143. 重排链表.cpp
#include <vector>

typedef struct ListNode {
    int val;
    ListNode* next;
    ListNode(int val=0, ListNode* next=nullptr) : val(val), next(next) {}
} ListNode;

class Solution {
public:
    // 递归法（这个方法需要返回值）
    ListNode* reorderList(ListNode* head) {
        if (head == nullptr || head->next == nullptr || head->next->next == nullptr) {
            return head;
        }
        // 找到当前非空尾节点和上一个节点
        ListNode *p = head, *q = head->next;
        while (q->next) {
            p = p->next;
            q = q->next;
        }
        // 重新排列
        p->next = nullptr;
        ListNode* newHead = head->next;
        head->next = q;
        q->next = reorderList(newHead);
        return head;
    }

    // 全部保存法
    void reorderList2(ListNode* head) {
        // 使用vector保存所有节点
        std::vector<ListNode*> stack;
        ListNode* tmp = head;
        while (tmp) {
            stack.push_back(tmp);
            tmp = tmp->next;
        }
        // 分别从前后取值，使用flag来区分
        int flag = 0;
        int l = 0, r = stack.size() - 1;
        ListNode* newHead = new ListNode(0);
        while (l <= r) {
            if (flag == 0) {
                newHead->next = stack[l++];
            } else {
                newHead->next = stack[r--];
            }
            flag = 1 - flag;
            newHead = newHead->next;
        }
        newHead->next = nullptr;
    }

    // 后一半保存，插入前一半
    void reorderList3(ListNode* head) {
        if (head == nullptr) {
            return;
        }
        // 找中点，断开
        ListNode *p = head, *q = head;
        while (p && q && q->next) {
            p = p->next;
            q = q->next->next;
        }
        // 保存后半部分到栈
        std::vector<ListNode*> stack;
        while (p) {
            stack.push_back(p);
            p = p->next;
        }
        // 从头到尾完成插入操作
        while (head && stack.size() > 1) {
            ListNode* tmp = head->next;
            head->next = stack.back();
            head->next->next = tmp;
            head = tmp;
            stack.pop_back();
            if (!stack.empty()) {
                stack.back()->next = nullptr;
            }
        }
    }

    // 找中点，翻转，归并
    void reorderList4(ListNode* head) {
        if (head == nullptr || head->next == nullptr) {
            return;
        }
        // 找中点，断开
        ListNode *mid = head, *p = head, *q = head;
        while (q && q->next) {
            mid = p;
            p = p->next;
            q = q->next->next;
        }
        mid->next = nullptr;
        // 后半部分翻转
        ListNode* newHead = nullptr;
        while (p) {
            ListNode* tmp = p->next;
            p->next = newHead;
            newHead = p;
            p = tmp;
        }
        // 两链表合并
        ListNode* h = new ListNode(0);
        while (head || newHead) {
            if (head) {
                h->next = head;
                h = h->next;
                head = head->next;
            }
            if (newHead) {
                h->next = newHead;
                h = h->next;
                newHead = newHead->next;
            }
        }
    }
};